<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="es">
<head>
<!-- Generated by javadoc (11.0.10) on Thu Apr 08 15:56:10 CEST 2021 -->
<title>FSSubsetIndividual (Ristretto API)</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2021-04-08">
<link rel="stylesheet" type="text/css" href="../../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../../jquery/jquery-ui.css" title="Style">
<script type="text/javascript" src="../../../../script.js"></script>
<script type="text/javascript" src="../../../../jquery/jszip/dist/jszip.min.js"></script>
<script type="text/javascript" src="../../../../jquery/jszip-utils/dist/jszip-utils.min.js"></script>
<!--[if IE]>
<script type="text/javascript" src="../../../../jquery/jszip-utils/dist/jszip-utils-ie.min.js"></script>
<![endif]-->
<script type="text/javascript" src="../../../../jquery/jquery-3.5.1.js"></script>
<script type="text/javascript" src="../../../../jquery/jquery-ui.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="FSSubsetIndividual (Ristretto API)";
        }
    }
    catch(err) {
    }
//-->
var data = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10,"i12":10,"i13":10,"i14":10,"i15":10,"i16":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
var pathtoroot = "../../../../";
var useModuleDirectories = true;
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<header role="banner">
<nav role="navigation">
<div class="fixedNav">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a id="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a id="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
<div class="aboutLanguage"><b>Ristretto API</div>
</div>
<div class="subNav">
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../allclasses.html">All&nbsp;Classes</a></li>
</ul>
<ul class="navListSearch">
<li><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a id="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
</div>
<div class="navPadding">&nbsp;</div>
<script type="text/javascript"><!--
$('.navPadding').css('padding-top', $('.fixedNav').css("height"));
//-->
</script>
</nav>
</header>
<!-- ======== START OF CLASS DATA ======== -->
<main role="main">
<div class="header">
<div class="subTitle"><span class="packageLabelInType">Package</span>&nbsp;<a href="package-summary.html">ristretto.problem.fs.subset</a></div>
<h2 title="Class FSSubsetIndividual" class="title">Class FSSubsetIndividual</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>ec.Individual</li>
<li>
<ul class="inheritance">
<li>ristretto.problem.fs.subset.FSSubsetIndividual</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><code>ec.Prototype</code>, <code>ec.Setup</code>, <code>java.io.Serializable</code>, <code>java.lang.Cloneable</code>, <code>java.lang.Comparable</code></dd>
</dl>
<hr>
<pre>public class <span class="typeNameLabel">FSSubsetIndividual</span>
extends ec.Individual</pre>
<div class="block">FSSubsetIndividual is an individual whose genome is a subset of selected
 features for a feature selection problem. As features are referenced by
 indexes, genomes will be subsets of integer indexes (starting from 0). It has
 been adapted from the original ECJ's IntegerVectorIndividual.java (Copyright
 2006 by Sean Luke under the
 <a href="https://opensource.org/licenses/AFL-3.0">AFL v. 3.0</a>) to keep the
 genome in a Javas's TreeSet.

 <p>
 This class contains two methods,
 <a href="#defaultCrossover(ec.EvolutionState,int,ristretto.problem.fs.subset.FSSubsetIndividual)"><code>defaultCrossover(ec.EvolutionState, int, ristretto.problem.fs.subset.FSSubsetIndividual)</code></a> and
 <a href="#defaultMutate(ec.EvolutionState,int)"><code>defaultMutate(ec.EvolutionState, int)</code></a>, which
 can be overridden if all you need is a simple crossover and a simple mutate
 mechanism. The
 <a href="breed/FSSubsetCrossoverPipeline.html" title="class in ristretto.problem.fs.subset.breed"><code>FSSubsetCrossoverPipeline</code></a> and
 <a href="breed/FSSubsetMutationPipeline.html" title="class in ristretto.problem.fs.subset.breed"><code>FSSubsetMutationPipeline</code></a> classes
 use these methods to do their handwork. For more sophisticated crossover and
 mutation, you'll need to write a custom breeding pipeline.

 <p>
 FSSubsetIndividuals must belong to the species
 <a href="FSSubsetSpecies.html" title="class in ristretto.problem.fs.subset"><code>FSSubsetSpecies</code></a> (or any subclass of it).

 <p>
 <b>From ec.Individual:</b>

 <p>
 In addition to serialization for checkpointing, Individuals may read and
 write themselves to streams in three ways.

 <ul>
 <li><b>writeIndividual(...,DataOutput)/readIndividual(...,DataInput)</b>&nbsp;&nbsp;&nbsp;This
 method transmits or receives an individual in binary. It is the most
 efficient approach to sending individuals over networks, etc. These methods
 write the evaluated flag and the fitness, then call
 <b>readGenotype/writeGenotype</b>, which you must implement to write those
 parts of your Individual special to your functions-- the default versions of
 readGenotype/writeGenotype throw errors. You don't need to implement them if
 you don't plan on using read/writeIndividual.

 <li><b>printIndividual(...,PrintWriter)/readIndividual(...,LineNumberReader)</b>&nbsp;&nbsp;&nbsp;This
 approach transmits or receives an individual in text encoded such that the
 individual is largely readable by humans but can be read back in 100% by ECJ
 as well. To do this, these methods will encode numbers using the
 <tt>ec.util.Code</tt> class. These methods are mostly used to write out
 populations to files for inspection, slight modification, then reading back
 in later on. <b>readIndividual</b> reads in the fitness and the evaluation
 flag, then calls <b>parseGenotype</b> to read in the remaining individual.
 You are responsible for implementing parseGenotype: the Code class is there
 to help you. <b>printIndividual</b> writes out the fitness and evaluation
 flag, then calls <b>genotypeToString</b> and printlns the resultant string.
 You are responsible for implementing the genotypeToString method in such a
 way that parseGenotype can read back in the individual println'd with
 genotypeToString. The default form of genotypeToString simply calls
 <b>toString</b>, which you may override instead if you like. The default form
 of <b>parseGenotype</b> throws an error. You are not required to implement
 these methods, but without them you will not be able to write individuals to
 files in a simultaneously computer- and human-readable fashion.

 <li><b>printIndividualForHumans(...,PrintWriter)</b>&nbsp;&nbsp;&nbsp;This
 approach prints an individual in a fashion intended for human consumption
 only. <b>printIndividualForHumans</b> writes out the fitness and evaluation
 flag, then calls <b>genotypeToStringForHumans</b> and printlns the resultant
 string. You are responsible for implementing the genotypeToStringForHumans
 method. The default form of genotypeToStringForHumans simply calls
 <b>toString</b>, which you may override instead if you like (though note that
 genotypeToString's default also calls toString). You should handle one of
 these methods properly to ensure individuals can be printed by ECJ.
 </ul>
 
 <p>
 In general, the various readers and writers do three things: they tell the
 Fitness to read/write itself, they read/write the evaluated flag, and they
 read/write the gene subset. If you add instance variables to a
 FSSubsetIndividual or subclass, you'll need to read/write those variables as
 well.</div>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../serialized-form.html#ristretto.problem.fs.subset.FSSubsetIndividual">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<section role="region">
<ul class="blockList">
<li class="blockList"><a id="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colSecond" scope="col">Field</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.TreeSet&lt;java.lang.Integer&gt;</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#genome">genome</a></span></code></th>
<td class="colLast">
<div class="block">Subset of selected feature indices</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#P_FS_SUBSET_INDIVIDUAL">P_FS_SUBSET_INDIVIDUAL</a></span></code></th>
<td class="colLast">
<div class="block">Base parameter</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a id="fields.inherited.from.class.ec.Individual">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;ec.Individual</h3>
<code>evaluated, EVALUATED_PREAMBLE, fitness, P_INDIVIDUAL, species</code></li>
</ul>
</li>
</ul>
</section>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<section role="region">
<ul class="blockList">
<li class="blockList"><a id="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Constructor</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tr class="altColor">
<th class="colConstructorName" scope="row"><code><span class="memberNameLink"><a href="#%3Cinit%3E()">FSSubsetIndividual</a></span>()</code></th>
<td class="colLast">&nbsp;</td>
</tr>
</table>
</li>
</ul>
</section>
<!-- ========== METHOD SUMMARY =========== -->
<section role="region">
<ul class="blockList">
<li class="blockList"><a id="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colSecond" scope="col">Method</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#clone()">clone</a></span>()</code></th>
<td class="colLast">
<div class="block">Returns a clone of this individual</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>ec.util.Parameter</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#defaultBase()">defaultBase</a></span>()</code></th>
<td class="colLast">
<div class="block">Return the default parameter base.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#defaultCrossover(ec.EvolutionState,int,ristretto.problem.fs.subset.FSSubsetIndividual)">defaultCrossover</a></span>&#8203;(ec.EvolutionState&nbsp;state,
                int&nbsp;thread,
                <a href="FSSubsetIndividual.html" title="class in ristretto.problem.fs.subset">FSSubsetIndividual</a>&nbsp;other)</code></th>
<td class="colLast">
<div class="block">Destructively crosses over the individual with another in some default
 manner.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#defaultMutate(ec.EvolutionState,int)">defaultMutate</a></span>&#8203;(ec.EvolutionState&nbsp;state,
             int&nbsp;thread)</code></th>
<td class="colLast">
<div class="block">Destructively mutates the individual in some default manner.</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#equals(java.lang.Object)">equals</a></span>&#8203;(java.lang.Object&nbsp;other)</code></th>
<td class="colLast">
<div class="block">Return true if this individual equals the other</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>int</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#genomeLength()">genomeLength</a></span>()</code></th>
<td class="colLast">
<div class="block">Return the length of the genome.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#genotypeToString()">genotypeToString</a></span>()</code></th>
<td class="colLast">
<div class="block">Convert the gnome into a string</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#genotypeToStringForHumans()">genotypeToStringForHumans</a></span>()</code></th>
<td class="colLast">
<div class="block">Return the gnome in a human-readable format</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#getGenome()">getGenome</a></span>()</code></th>
<td class="colLast">
<div class="block">Return the genome.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>int</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#hashCode()">hashCode</a></span>()</code></th>
<td class="colLast">
<div class="block">Return the hascode of this individual</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>protected void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#parseGenotype(ec.EvolutionState,java.io.LineNumberReader)">parseGenotype</a></span>&#8203;(ec.EvolutionState&nbsp;state,
             java.io.LineNumberReader&nbsp;reader)</code></th>
<td class="colLast">
<div class="block">Parse a genome from a numbered reader.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#readGenotype(ec.EvolutionState,java.io.DataInput)">readGenotype</a></span>&#8203;(ec.EvolutionState&nbsp;state,
            java.io.DataInput&nbsp;dataInput)</code></th>
<td class="colLast">
<div class="block">Read the genome from a binary stream.</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#reset(ec.EvolutionState,int)">reset</a></span>&#8203;(ec.EvolutionState&nbsp;state,
     int&nbsp;thread)</code></th>
<td class="colLast">
<div class="block">Initializes the individual by randomly choosing features.</div>
</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#setGenome(java.lang.Object)">setGenome</a></span>&#8203;(java.lang.Object&nbsp;gen)</code></th>
<td class="colLast">
<div class="block">Set the genome.</div>
</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></span>&#8203;(ec.EvolutionState&nbsp;state,
     ec.util.Parameter&nbsp;base)</code></th>
<td class="colLast">
<div class="block">Set up this individual.</div>
</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code>long</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#size()">size</a></span>()</code></th>
<td class="colLast">
<div class="block">Return the number of selected features</div>
</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>void</code></td>
<th class="colSecond" scope="row"><code><span class="memberNameLink"><a href="#writeGenotype(ec.EvolutionState,java.io.DataOutput)">writeGenotype</a></span>&#8203;(ec.EvolutionState&nbsp;state,
             java.io.DataOutput&nbsp;dataOutput)</code></th>
<td class="colLast">
<div class="block">Write the genome to a binary stream.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a id="methods.inherited.from.class.ec.Individual">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.Individual</h3>
<code>compareTo, distanceTo, merge, printIndividual, printIndividual, printIndividual, printIndividualForHumans, printIndividualForHumans, readIndividual, readIndividual, toString, writeIndividual</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a id="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>finalize, getClass, notify, notifyAll, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</section>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<section role="region">
<ul class="blockList">
<li class="blockList"><a id="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a id="P_FS_SUBSET_INDIVIDUAL">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_FS_SUBSET_INDIVIDUAL</h4>
<pre>public static final&nbsp;java.lang.String P_FS_SUBSET_INDIVIDUAL</pre>
<div class="block">Base parameter</div>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../constant-values.html#ristretto.problem.fs.subset.FSSubsetIndividual.P_FS_SUBSET_INDIVIDUAL">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a id="genome">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>genome</h4>
<pre>public&nbsp;java.util.TreeSet&lt;java.lang.Integer&gt; genome</pre>
<div class="block">Subset of selected feature indices</div>
</li>
</ul>
</li>
</ul>
</section>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<section role="region">
<ul class="blockList">
<li class="blockList"><a id="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a id="&lt;init&gt;()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>FSSubsetIndividual</h4>
<pre>public&nbsp;FSSubsetIndividual()</pre>
</li>
</ul>
</li>
</ul>
</section>
<!-- ============ METHOD DETAIL ========== -->
<section role="region">
<ul class="blockList">
<li class="blockList"><a id="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a id="defaultBase()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultBase</h4>
<pre class="methodSignature">public&nbsp;ec.util.Parameter&nbsp;defaultBase()</pre>
<div class="block">Return the default parameter base.</div>
</li>
</ul>
<a id="clone()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clone</h4>
<pre class="methodSignature">public&nbsp;java.lang.Object&nbsp;clone()</pre>
<div class="block">Returns a clone of this individual</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code>clone</code>&nbsp;in interface&nbsp;<code>ec.Prototype</code></dd>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>clone</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
</dl>
</li>
</ul>
<a id="setup(ec.EvolutionState,ec.util.Parameter)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setup</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;setup&#8203;(ec.EvolutionState&nbsp;state,
                  ec.util.Parameter&nbsp;base)</pre>
<div class="block">Set up this individual.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code>setup</code>&nbsp;in interface&nbsp;<code>ec.Prototype</code></dd>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code>setup</code>&nbsp;in interface&nbsp;<code>ec.Setup</code></dd>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>setup</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>base</code> - The parameter base</dd>
</dl>
</li>
</ul>
<a id="defaultCrossover(ec.EvolutionState,int,ristretto.problem.fs.subset.FSSubsetIndividual)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultCrossover</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;defaultCrossover&#8203;(ec.EvolutionState&nbsp;state,
                             int&nbsp;thread,
                             <a href="FSSubsetIndividual.html" title="class in ristretto.problem.fs.subset">FSSubsetIndividual</a>&nbsp;other)</pre>
<div class="block">Destructively crosses over the individual with another in some default
 manner. The default form preserves the common selected features in both
 individuals and exchanges a random number of feature indexes.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>thread</code> - Execution thread</dd>
<dd><code>other</code> - The other individual</dd>
</dl>
</li>
</ul>
<a id="defaultMutate(ec.EvolutionState,int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultMutate</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;defaultMutate&#8203;(ec.EvolutionState&nbsp;state,
                          int&nbsp;thread)</pre>
<div class="block">Destructively mutates the individual in some default manner.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>thread</code> - Execution thread</dd>
</dl>
</li>
</ul>
<a id="reset(ec.EvolutionState,int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>reset</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;reset&#8203;(ec.EvolutionState&nbsp;state,
                  int&nbsp;thread)</pre>
<div class="block">Initializes the individual by randomly choosing features.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>thread</code> - Execution thread</dd>
</dl>
</li>
</ul>
<a id="hashCode()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>hashCode</h4>
<pre class="methodSignature">public&nbsp;int&nbsp;hashCode()</pre>
<div class="block">Return the hascode of this individual</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code>hashCode</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
</dl>
</li>
</ul>
<a id="genotypeToStringForHumans()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genotypeToStringForHumans</h4>
<pre class="methodSignature">public&nbsp;java.lang.String&nbsp;genotypeToStringForHumans()</pre>
<div class="block">Return the gnome in a human-readable format</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>genotypeToStringForHumans</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
</dl>
</li>
</ul>
<a id="genotypeToString()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genotypeToString</h4>
<pre class="methodSignature">public&nbsp;java.lang.String&nbsp;genotypeToString()</pre>
<div class="block">Convert the gnome into a string</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>genotypeToString</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
</dl>
</li>
</ul>
<a id="parseGenotype(ec.EvolutionState,java.io.LineNumberReader)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseGenotype</h4>
<pre class="methodSignature">protected&nbsp;void&nbsp;parseGenotype&#8203;(ec.EvolutionState&nbsp;state,
                             java.io.LineNumberReader&nbsp;reader)
                      throws java.io.IOException</pre>
<div class="block">Parse a genome from a numbered reader.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>parseGenotype</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>reader</code> - The reader</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>java.io.IOException</code> - if the genome can not be parsed</dd>
</dl>
</li>
</ul>
<a id="equals(java.lang.Object)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>equals</h4>
<pre class="methodSignature">public&nbsp;boolean&nbsp;equals&#8203;(java.lang.Object&nbsp;other)</pre>
<div class="block">Return true if this individual equals the other</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code>equals</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>other</code> - The other individual</dd>
</dl>
</li>
</ul>
<a id="getGenome()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getGenome</h4>
<pre class="methodSignature">public&nbsp;java.lang.Object&nbsp;getGenome()</pre>
<div class="block">Return the genome.</div>
</li>
</ul>
<a id="setGenome(java.lang.Object)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setGenome</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;setGenome&#8203;(java.lang.Object&nbsp;gen)</pre>
<div class="block">Set the genome.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>gen</code> - The new genoms</dd>
</dl>
</li>
</ul>
<a id="genomeLength()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genomeLength</h4>
<pre class="methodSignature">public&nbsp;int&nbsp;genomeLength()</pre>
<div class="block">Return the length of the genome.</div>
</li>
</ul>
<a id="writeGenotype(ec.EvolutionState,java.io.DataOutput)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writeGenotype</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;writeGenotype&#8203;(ec.EvolutionState&nbsp;state,
                          java.io.DataOutput&nbsp;dataOutput)
                   throws java.io.IOException</pre>
<div class="block">Write the genome to a binary stream.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>writeGenotype</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>dataOutput</code> - The stream writer</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>java.io.IOException</code> - If the genome can not be written</dd>
</dl>
</li>
</ul>
<a id="readGenotype(ec.EvolutionState,java.io.DataInput)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readGenotype</h4>
<pre class="methodSignature">public&nbsp;void&nbsp;readGenotype&#8203;(ec.EvolutionState&nbsp;state,
                         java.io.DataInput&nbsp;dataInput)
                  throws java.io.IOException</pre>
<div class="block">Read the genome from a binary stream.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>readGenotype</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>state</code> - The evolution state</dd>
<dd><code>dataInput</code> - The stream reader</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>java.io.IOException</code> - If the genome can not be read</dd>
</dl>
</li>
</ul>
<a id="size()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>size</h4>
<pre class="methodSignature">public&nbsp;long&nbsp;size()</pre>
<div class="block">Return the number of selected features</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>size</code>&nbsp;in class&nbsp;<code>ec.Individual</code></dd>
</dl>
</li>
</ul>
</li>
</ul>
</section>
</li>
</ul>
</div>
</div>
</main>
<!-- ========= END OF CLASS DATA ========= -->
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a id="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a id="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
<div class="aboutLanguage"><b>Ristretto API</div>
</div>
<div class="subNav">
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../allclasses.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a id="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</nav>
<p class="legalCopy"><small><font size="-1"><div align=right> Copyright &copy; 2016-2018 <a href="https://atcproyectos.ugr.es/efficomp/">EFFICOMP</a>. All rights reserved. <br><br> </div><div align=justify> This work was supported by project <i>Energy-aware High Performance Multi-objective Optimization in Heterogeneous Computer Architectures. Applications on Biomedical Engineering</i> (<a href="https://atcproyectos.ugr.es/ehpmobe/">e-hpMOBE</a>), with reference TIN2015-67020-P, funded by the Spanish <i> <a href="https://www.ciencia.gob.es/">Ministerio de Economía y Competitividad</a></i>, and by the <a href="https://ec.europa.eu/regional_policy/en/funding/erdf/"> European Regional Development Fund (ERDF)</a>. <br> <br> </div><div align=center> <a href="https://www.ciencia.gob.es/"><img height="75" src="https://raw.githubusercontent.com/efficomp/ristretto/main/docs/resources/mineco.png" alt="Ministerio de Economía y Competitividad"></a> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <a href="https://ec.europa.eu/regional_policy/en/funding/erdf/"> <img height="75" src="https://raw.githubusercontent.com/efficomp/ristretto/main/docs/resources/erdf.png" alt="European Regional Development Fund (ERDF)"></a> </div></font></small></p>
</footer>
</body>
</html>
